File: backend/managers/profileManager.js
const path = require('path');
const fs = require('fs');
const { v4: uuidv4 } = require('uuid');
const { 
  loadConfig, 
  saveConfig, 
  getModsPath 
} = require('../core/config');

// We'll lazy-load modManager to avoid circular dependencies if possible, 
// or carefully structure our imports. 
// For now, we might need to access mod paths directly or use helper functions.

class ProfileManager {
  constructor() {
    this.initialized = false;
  }

  init() {
    if (this.initialized) return;

    const config = loadConfig();
    
    // Migration: specific check to see if we have profiles yet
    if (!config.profiles || Object.keys(config.profiles).length === 0) {
      this.migrateLegacyConfig(config);
    }
    
    this.initialized = true;
    console.log('[ProfileManager] Initialized');
  }

  migrateLegacyConfig(config) {
    console.log('[ProfileManager] Migrating legacy config to profile system...');
    
    // Create a default profile with current settings
    const defaultProfileId = 'default';
    const defaultProfile = {
      id: defaultProfileId,
      name: 'Default',
      created: new Date().toISOString(),
      lastUsed: new Date().toISOString(),
      
      // settings specific to this profile
      // If global settings existed, we copy them here
      mods: config.installedMods || [], // Legacy mods are now part of default profile
      javaPath: config.javaPath || '',
      gameOptions: {
        minMemory: '1G',
        maxMemory: '4G',
        args: []
      }
    };

    const updates = {
      profiles: {
        [defaultProfileId]: defaultProfile
      },
      activeProfileId: defaultProfileId,
      // We keep a global registry of "known" mods if we want, 
      // but for now the current implementation implies 
      // mods are just files in folders.
      // We'll use the profile "mods" array to track ENABLED/KNOWN mods for that profile.
    };

    saveConfig(updates);
    console.log('[ProfileManager] Migration complete. Created Default profile.');
  }

  createProfile(name) {
    if (!name || typeof name !== 'string') {
      throw new Error('Invalid profile name');
    }

    const config = loadConfig();
    const id = uuidv4();
    
    const newProfile = {
      id,
      name: name.trim(),
      created: new Date().toISOString(),
      lastUsed: null,
      mods: [], // Start with no mods enabled
      javaPath: '', 
      gameOptions: {
        minMemory: '1G',
        maxMemory: '4G',
        args: []
      }
    };

    const profiles = config.profiles || {};
    profiles[id] = newProfile;

    saveConfig({ profiles });
    
    console.log(`[ProfileManager] Created new profile: "${name}" (${id})`);
    return newProfile;
  }

  getProfiles() {
    const config = loadConfig();
    return Object.values(config.profiles || {});
  }

  getProfile(id) {
    const config = loadConfig();
    return (config.profiles && config.profiles[id]) || null;
  }

  getActiveProfile() {
    const config = loadConfig();
    const activeId = config.activeProfileId;
    if (!activeId || !config.profiles || !config.profiles[activeId]) {
      // Fallback if something is corrupted
      return this.getProfiles()[0] || null; 
    }
    return config.profiles[activeId];
  }

  async activateProfile(id) {
    const config = loadConfig();
    if (!config.profiles || !config.profiles[id]) {
        throw new Error(`Profile not found: ${id}`);
    }

    if (config.activeProfileId === id) {
        console.log(`[ProfileManager] Profile ${id} is already active.`);
        return config.profiles[id];
    }

    console.log(`[ProfileManager] Switching to profile: ${config.profiles[id].name} (${id})`);
    
    // 1. Update config first
    config.profiles[id].lastUsed = new Date().toISOString();
    saveConfig({ 
        activeProfileId: id,
        profiles: config.profiles 
    });

    // 2. Trigger Mod Sync
    // We need to require this here to ensure it uses the *newly saved* active profile ID
    const { syncModsForCurrentProfile } = require('./modManager');
    await syncModsForCurrentProfile();

    return config.profiles[id];
  }

  deleteProfile(id) {
    const config = loadConfig();
    const profiles = config.profiles || {};
    
    if (!profiles[id]) {
      throw new Error('Profile not found');
    }

    if (config.activeProfileId === id) {
       throw new Error('Cannot delete the active profile');
    }

    // Don't allow deleting the last profile
    if (Object.keys(profiles).length <= 1) {
        throw new Error('Cannot delete the only remaining profile');
    }

    delete profiles[id];
    saveConfig({ profiles });
    console.log(`[ProfileManager] Deleted profile: ${id}`);
    
    return true;
  }

  updateProfile(id, updates) {
    const config = loadConfig();
    const profiles = config.profiles || {};
    
    if (!profiles[id]) {
        throw new Error('Profile not found');
    }

    // Safety checks on updates
    const allowedFields = ['name', 'javaPath', 'gameOptions', 'mods'];
    const sanitizedUpdates = {};
    
    Object.keys(updates).forEach(key => {
        if (allowedFields.includes(key)) {
            sanitizedUpdates[key] = updates[key];
        }
    });

    profiles[id] = { ...profiles[id], ...sanitizedUpdates };
    
    saveConfig({ profiles });
    console.log(`[ProfileManager] Updated profile: ${id}`);
    
    // If we updated mods for the *active* profile, we might need to sync immediately
    if (config.activeProfileId === id && updates.mods) {
         // Optionally trigger sync? 
         // Usually updates come from "Enabling/Disabling" a single mod, 
         // which might call a more specific method. 
         // But if we bulk update, we should sync.
         // Let's leave sync invoke to the caller or specific methods for now.
    }

    return profiles[id];
  }
}

module.exports = new ProfileManager();

File: lib/core/managers/profile_manager.dart
import 'package:uuid/uuid.dart';
import '../config_manager.dart';

class ProfileManager {
  static bool _initialized = false;

  static Future<void> init() async {
    if (_initialized) return;
    final config = await ConfigManager().loadConfig();
    final profiles = config['profiles'];
    if (profiles == null || (profiles is Map && profiles.isEmpty)) {
      await _migrateLegacyConfig(config);
    }
    _initialized = true;
  }

  static Future<void> _migrateLegacyConfig(Map<String, dynamic> config) async {
    const defaultProfileId = "default";
    final now = DateTime.now().toUtc().toIso8601String();
    final mods = config['installedMods'] ?? [];
    final javaPath = config['javaPath'] ?? "";
    
    final defaultProfile = {
      "id": defaultProfileId,
      "name": "Default",
      "created": now,
      "lastUsed": now,
      "mods": mods,
      "javaPath": javaPath,
      "gameOptions": {
        "minMemory": "1G",
        "maxMemory": "4G",
        "args": []
      }
    };
    
    final newConfig = Map<String, dynamic>.from(config);
    newConfig['profiles'] = {defaultProfileId: defaultProfile};
    newConfig['activeProfileId'] = defaultProfileId;
    
    await ConfigManager().saveConfigObject(newConfig);
  }

  static Future<Map<String, dynamic>> createProfile(String name) async {
    if (name.trim().isEmpty) {
      throw Exception("Name cannot be blank");
    }
    final config = await ConfigManager().loadConfig();
    final profiles = Map<String, dynamic>.from(config['profiles'] ?? {});
    final profileId = const Uuid().v4();
    final now = DateTime.now().toUtc().toIso8601String();
    
    final profile = {
      "id": profileId,
      "name": name.trim(),
      "created": now,
      "lastUsed": now,
      "mods": [],
      "javaPath": "",
      "gameOptions": {
        "minMemory": "1G",
        "maxMemory": "4G",
        "args": []
      }
    };
    
    profiles[profileId] = profile;
    config['profiles'] = profiles;
    await ConfigManager().saveConfigObject(config);
    return profile;
  }

  static Future<List<Map<String, dynamic>>> getProfiles() async {
    final config = await ConfigManager().loadConfig();
    final profiles = config['profiles'];
    if (profiles is Map) {
      return profiles.values.map((e) => Map<String, dynamic>.from(e)).toList();
    }
    return [];
  }

  static Future<Map<String, dynamic>?> getProfile(String profileId) async {
    final config = await ConfigManager().loadConfig();
    final profiles = config['profiles'];
    if (profiles is Map) {
      final profile = profiles[profileId];
      if (profile != null) {
        return Map<String, dynamic>.from(profile);
      }
    }
    return null;
  }

  static Future<Map<String, dynamic>?> getActiveProfile() async {
    final config = await ConfigManager().loadConfig();
    final activeId = config['activeProfileId'];
    final profiles = config['profiles'];
    
    if (profiles == null || (profiles is Map && profiles.isEmpty)) {
      return null;
    }
    
    if (activeId != null && profiles is Map && profiles.containsKey(activeId)) {
      return Map<String, dynamic>.from(profiles[activeId]);
    }
    
    if (profiles is Map && profiles.isNotEmpty) {
      return Map<String, dynamic>.from(profiles.values.first);
    }
    return null;
  }

  static Future<Map<String, dynamic>> activateProfile(String profileId) async {
    final config = await ConfigManager().loadConfig();
    final profiles = Map<String, dynamic>.from(config['profiles'] ?? {});
    
    if (!profiles.containsKey(profileId)) {
      throw Exception("Profile not found");
    }
    
    final profile = Map<String, dynamic>.from(profiles[profileId]);
    profile['lastUsed'] = DateTime.now().toUtc().toIso8601String();
    profiles[profileId] = profile;
    
    config['profiles'] = profiles;
    config['activeProfileId'] = profileId;
    await ConfigManager().saveConfigObject(config);
    return profile;
  }

  static Future<bool> deleteProfile(String profileId) async {
    final config = await ConfigManager().loadConfig();
    final profiles = Map<String, dynamic>.from(config['profiles'] ?? {});
    final activeId = config['activeProfileId'];
    
    if (activeId == profileId) {
      throw Exception("Cannot delete the active profile");
    }
    if (profiles.length <= 1) {
      throw Exception("Cannot delete the only remaining profile");
    }
    if (!profiles.containsKey(profileId)) {
      return false;
    }
    
    profiles.remove(profileId);
    config['profiles'] = profiles;
    await ConfigManager().saveConfigObject(config);
    return true;
  }

  static Future<Map<String, dynamic>> updateProfile(String profileId, Map<String, dynamic> updates) async {
    final config = await ConfigManager().loadConfig();
    final profiles = Map<String, dynamic>.from(config['profiles'] ?? {});
    
    if (!profiles.containsKey(profileId)) {
      throw Exception("Profile not found");
    }
    
    final profile = Map<String, dynamic>.from(profiles[profileId]);
    final allowed = {"name", "javaPath", "gameOptions", "mods"};
    
    for (final entry in updates.entries) {
      if (allowed.contains(entry.key)) {
        profile[entry.key] = entry.value;
      }
    }
    
    profiles[profileId] = profile;
    config['profiles'] = profiles;
    await ConfigManager().saveConfigObject(config);
    return profile;
  }
}
